### PR TITLE
RTL872x: Implement HAL event group for UART

### PR DESCRIPTION
**Pull request by XuGuohui**
### Problem
For platforms, like TrackerM, rely on the UART HAL event group APIs, but they are not implemented yet.

### Solution
Implements the UART HAL event group APIs. This PR also implements the UART HAL sleep API.

### Example App
 This test app is for verifying the HAL sleep API.
```c
#include "Particle.h"

SYSTEM_MODE(MANUAL);
SYSTEM_THREAD(ENABLED);

#define SERIAL                  Serial2
#define SERIAL_INTERFACE        HAL_USART_SERIAL2

void setup() {
    SERIAL.begin(115200);
    pinMode(D11, INPUT_PULLUP);
}

void loop() {
    if (digitalRead(D11) == LOW) {
        delay(1s);
        SERIAL.printlnf("available1: %d", SERIAL.available());
        hal_usart_sleep(SERIAL_INTERFACE, true, nullptr);
        while (digitalRead(D11) == LOW);
        hal_usart_sleep(SERIAL_INTERFACE, false, nullptr);
        delay(1s);
        SERIAL.printlnf("available2: %d", SERIAL.available());
        while (SERIAL.available()) {
            SERIAL.write(SERIAL.read());
        }
        SERIAL.println(" ");
    }

    SERIAL.println("This is loop()");
    delay(1s);
}
```

### References

N/A

---

### Completeness

- [x] User is totes amazing for contributing!
- [x] Contributor has signed CLA ([Info here](https://github.com/spark/firmware/blob/develop/CONTRIBUTING.md))
- [ ] Problem and Solution clearly stated
- [ ] Run unit/integration/application tests on device
- [ ] Added documentation
- [ ] Added to CHANGELOG.md after merging (add links to docs and issues)

### CODE REVIEW

File: hal/src/rtl872x/usart_hal.cpp [hunk 1]
```diff
@@ -717,6 +826,10 @@ class Usart {
         uart->dataInFlight(true /* commit */, true /* cancel */);
         uart->receiving_ = false;
         uart->startReceiver();
+        BaseType_t yield = pdFALSE;
+        if (xEventGroupSetBitsFromISR(uart->evGroup_, HAL_USART_PVT_EVENT_READABLE, &yield) != pdFAIL) {
```
Comment (avtolstoy) [inline hal/src/rtl872x/usart_hal.cpp:830]: This doesn't look right. This interrupt handler only generates at the end of the full DMA transfer. This event should be generated once we receive even just 1 byte.

Comment (XuGuohui) [reply hal/src/rtl872x/usart_hal.cpp:830]: In that case, we need to enable rx interrupt when using  DMA mode and only set this event bit in the ISR. Sounds good?

Comment (avtolstoy) [reply hal/src/rtl872x/usart_hal.cpp:830]: Yes, that's the idea (it's the same on Gen 3). And disable immediately once it happens. Also this interrupt should ideally be triggered on transfer out of UART FIFO not on when the byte appears in UART FIFO. + We need to syncup the data in flight on getting unblocked by this event.

File: hal/src/rtl872x/usart_hal.cpp [hunk 2]
```diff
@@ -437,9 +440,7 @@ class Usart {
 
     int disableEvent(HAL_USART_Pvt_Events event) {
         if (event & HAL_USART_PVT_EVENT_READABLE) {
-            RxLock lk(this);
-            auto uartInstance = uartTable_[index_].UARTx;
-            uartInstance->MISCR &= ~(RUART_RXDMA_OWNER);
+            rxLock(true);
```
Comment (avtolstoy) [inline hal/src/rtl872x/usart_hal.cpp:443]: This is not correct. This should acquire RX lock, and
```c++
UART_INTConfig(uartInstance, (RUART_IER_ERBI | RUART_IER_ELSI | RUART_IER_ETOI), DISABLE);
UART_RXDMACmd(uartInstance, ENABLE);
```

**Review by avtolstoy — CHANGES_REQUESTED** [review]

**Review by keeramis — APPROVED** [review]
Verified on test runner and through user app that the new event group implementation is working 🎉

**Review by avtolstoy — CHANGES_REQUESTED** [review]

File: hal/src/rtl872x/usart_hal.cpp [hunk 3]
```diff
@@ -412,6 +408,68 @@ class Usart {
...
+                UART_RXDMACmd(uartInstance, DISABLE);
+            } else {
+                xEventGroupSetBits(evGroup_, HAL_USART_PVT_EVENT_READABLE);
+            }
+        }
+
+        if (event & HAL_USART_PVT_EVENT_WRITABLE) {
+            if (space() <= 0) {
+                TxLock lk(this);
+                UART_INTConfig(uartInstance, RUART_IER_ETBEI, ENABLE);
```
[the remark this answers is not in the document]
Comment (XuGuohui) [reply hal/src/rtl872x/usart_hal.cpp:432]: Needs to check `receiving_` as well?

**Review by avtolstoy — APPROVED** [review]

**Pull request merged**